twenty-first/src/shared_math/b_field_element.rs

```diff
@@ -269,6 +253,12 @@ impl From<u32> for BFieldElement {
     }
 }
 
+impl From<u64> for BFieldElement {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
```

Comment:
This behavior is unexpected.

From the [`std::convert::From` documentation](https://doc.rust-lang.org/std/convert/trait.From.html):

> **Note: This trait must not fail.** The From trait is intended for perfect conversions. If the conversion can fail or is not perfect, use [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html).

It is not generally true that `u64`s convert to `BFieldElement`s, since any `u64` $\geq P\not\in$ `BFieldElement`. 

By claiming that e.g. `BFieldElement::QUOTIENT.into() == BFieldElement::zero()` is expectable, this would justify `impl From<u64> for u32` or any other lossy conversion as implicitly modular. I think there is risk associated with making lossy conversion the default, since it might encourage (either accidentally or unknowingly) a conversion to happen when the caller does not assume that their `u64` contains a non-canonical `BFieldElement` value.

Reply:
Adding context without arguing (my opinion is not qualified enough yet):

I added `From<u64> for BFieldElement` even though `BFieldElement::new` already takes u64s: `u64 -> BFieldElement`. The rationale was that `new` does the modular reduction if necessary. But sometimes, like when reading hardcoded constants from the program code that are known to be BFieldElements and just happen to be encoded as u64s, you want to create the BFieldElement object directly and bypass the validity test and possible modular reduction. The way I read the documentation was that `from<u64> for BFieldElement` was the right way to do that.